Refactor Kata Gilde Rose Curso Clean Code Marcos Barreiro Otero

Mbarote opened the pull request:
1. Create a new class to list the item names in constants
2. Replace the for with a foreach to avoid repeating items[i]
3. Create a helper class with base functions to update sellIn and quality
4. Use a switch to process each item type in a separete function

Remaining work:

1. Add functions to the remaining items type

Mbarote committed:
Feat: initial refactor to updateQualityV2

Mbarote committed:
Feat: added Backstage

hectorcanto commented on Java/src/main/java/com/gildedrose/ItemNameValues.java:
```diff
@@ -0,0 +1,13 @@
+package com.gildedrose;
```
gústame que moveras as constantes fora e que reduciras a lonxitude dalguns nomes

hectorcanto commented on Java/src/main/java/com/gildedrose/GildedRose.java:
```diff
@@ -1,62 +1,115 @@
+        increaseBackStagePassesPostSellIn(item);
+    }
+
+    private void processSulfuras(Item item){
```
xenial!

hectorcanto commented on Java/src/main/java/com/gildedrose/GildedRose.java:
```diff
@@ -1,62 +1,115 @@
+        for (Item item : items) {
+            final String itemName = item.name;
+
+            switch (itemName){
```
aínda que aos Clean Coders máis puros non lles gusta o switch, o mellora é moi grande, moi ben.

Mbarote replied:
Engadino porque me parece que queda o código máis sinxelo de entender.

De todas formas para engadir o novo tipo de arquivo teño que cambiar a lóxica xa que agora teño que comprobar se o nome comeza cun determinado texto

hectorcanto commented on Java/src/main/java/com/gildedrose/GildedRose.java:
```diff
@@ -1,62 +1,115 @@
+    private void processBackstagePasses(Item item){
+        UpdateQualityHelper.increaseItemQuality(item, 1);
+        UpdateQualityHelper.updateSellIn(item);
+        increaseBackStagePassesPostSellIn(item);
```
preocúpame que o Helper non sexa capaz de resolver este último tamén, queda a lóxica rota en dous. Porén, si que o vin rápido así que visible é

hectorcanto commented on Java/src/main/java/com/gildedrose/GildedRose.java:
```diff
@@ -1,62 +1,115 @@
         this.items = items;
     }
 
+    public void updateQualityV2(){
```
Está ben isto do v2, entendo que cando acabes borraras o v1 e renomearas isto. Unha opción interesante é ter un wrapper que chame a v2 se está implementado e a v1 para o resto. De maneira que xa poderías lanzar os tests como están ou mesmo despregar a produción.

hectorcanto commented on Java/src/main/java/com/gildedrose/UpdateQualityHelper.java:
```diff
@@ -0,0 +1,29 @@
+
+import static com.gildedrose.ItemNameValues.SULFURAS;
+
+public class UpdateQualityHelper {
```
helper é un termo controvertido (como utils, aux, etc). Só o digo, neste contexto funciona

hectorcanto reviewed (changes requested):
a mellora é sustancial, case sería unha iteración completa

Mbarote committed:
Feature: added Conjured

Mbarote commented:
New changes:

1. Added rest of the items
2. Added logic for conjured items
3. Increase code for each item type

Remaining work:

1. Increase amount of tests to cover all items characteristics

Mbarote committed:
Feature: test Conjured
